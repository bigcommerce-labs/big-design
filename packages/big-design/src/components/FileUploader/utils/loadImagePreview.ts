export const loadImagePreview = (file: File): Promise<string | null> => {
  if (!file.type.startsWith('image/')) {
    return Promise.resolve(null);
  }

  return new Promise((resolve, reject) => {
    const reader = new FileReader();

    reader.onload = (event) => {
      // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
      const imageUrl = event.target?.result as string;

      resolve(imageUrl);
    };

    reader.onerror = (error) => {
      // eslint-disable-next-line @typescript-eslint/prefer-promise-reject-errors
      reject(error);
    };

    reader.readAsDataURL(file);
  });
};

export const getImagesPreview = async (files: File[]) =>
  files.reduce(async (prevPromise, file, idx) => {
    const prevResult = await prevPromise;
    const imageUrl = await loadImagePreview(file);

    return { ...prevResult, [idx]: imageUrl };
  }, Promise.resolve<Record<number, string | null>>({}));
