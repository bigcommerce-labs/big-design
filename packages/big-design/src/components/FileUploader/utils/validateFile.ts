import { ValidatorConfig } from '../types';

export const validateFileFormat = (file: File | DataTransferItem, fileFormats = '') => {
  const allowedFormats = fileFormats.split(',').map((format) => format.trim());
  const fileExtension = file.type.split('/').pop()?.toLowerCase() ?? '';

  return allowedFormats.some((format) => {
    if (format === '*' || format === '.*') {
      return true;
    } else if (format.startsWith('.')) {
      return format === `.${fileExtension}`;
    } else if (
      format.startsWith('audio/') ||
      format.startsWith('video/') ||
      format.startsWith('image/')
    ) {
      const mediaTypePrefix = format.split('/')[0];

      return file.type.startsWith(`${mediaTypePrefix}/`);
    }

    return true;
  });
};

export const validateFiles = (files: File[], validators: ValidatorConfig[]) =>
  files.flatMap((file, idx) =>
    validators
      .filter(({ validator }) => !validator(file))
      .map(({ message, type }) => ({ file, fileIdx: idx, message, type })),
  );
