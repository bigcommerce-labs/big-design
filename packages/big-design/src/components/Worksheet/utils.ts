import { Cell, WorksheetError, WorksheetItem } from './types';

export const mergeCells = <T extends WorksheetItem>(oldCells: Cell<T>[], newCells: Cell<T>[]) =>
  newCells.reduce(
    (accum, newCell) => {
      const index = oldCells.findIndex(
        (oldCell) => oldCell.rowIndex === newCell.rowIndex && oldCell.columnIndex === newCell.columnIndex,
      );

      if (index > -1) {
        accum[index] = newCell;

        return accum;
      }

      return accum.concat(newCell);
    },
    [...oldCells], // Note: returns a new array every time
  );

export const deleteCells = <T extends WorksheetItem>(oldCells: Cell<T>[], newCells: Cell<T>[]) =>
  oldCells.filter(
    (oldCell) =>
      !newCells.find((newCell) => newCell.columnIndex === oldCell.columnIndex && newCell.rowIndex === oldCell.rowIndex),
  );

export const editedRows = <T extends WorksheetItem>(editedCells: Cell<T>[], rows: T[]) =>
  editedCells.reduce<T[]>((accum, { rowIndex }) => {
    const row = rows[rowIndex];

    // Check to see if the row already exists in accum
    if (accum.find((editedRow) => editedRow === row)) {
      return accum;
    } else {
      // Only append new rows
      return [...accum, row];
    }
  }, []);

export const invalidRows = <T extends WorksheetItem>(invalidCells: Cell<T>[], rows: T[]) => {
  const mapObj = new Map();

  // Create Map with each row and append errors per row
  invalidCells.forEach(({ rowIndex, hash }) => {
    const row = rows[rowIndex];
    if (mapObj.has(row)) {
      const errors = mapObj.get(row);
      mapObj.set(row, new Set([...errors, hash]));
    } else {
      mapObj.set(row, new Set([hash]));
    }
  });

  return Array.from(mapObj).map<WorksheetError<T>>(([item, errors]) => ({
    item,
    errors: Array.from(errors),
  }));
};
