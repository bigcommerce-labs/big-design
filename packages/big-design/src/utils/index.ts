export * from './typedMemo';
export * from './uniqueId';
export * from './warning';
