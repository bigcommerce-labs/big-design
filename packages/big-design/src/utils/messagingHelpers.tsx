import { CheckCircleIcon, ErrorIcon, InfoIcon, WarningIcon } from '@bigcommerce/big-design-icons';
import { ThemeInterface } from '@bigcommerce/big-design-theme';
import React, { ComponentPropsWithoutRef } from 'react';
import { css } from 'styled-components';

import { ButtonProps } from '../components/Button';
import { LinkProps } from '../components/Link';
import { MarginProps } from '../mixins/margins';

export interface SharedMessagingProps extends ComponentPropsWithoutRef<'div'> {
  actions?: MessageAction[];
  header?: string;
  messages: MessageItem[];
  type?: MessagingType;
  onClose?(): void;
}

export type MessagingType = 'success' | 'error' | 'warning' | 'info';

export interface MessageItem {
  text: string;
  link?: MessageLinkItem;
}

export type MessageLinkItem = Pick<LinkProps, 'external' | 'href' | 'target'> & {
  text: string;
};

export interface MessageAction
  extends Omit<ButtonProps, 'children' | 'mobileWidth' | 'variant' | keyof MarginProps> {
  text: string;
  variant?: Exclude<ButtonProps['variant'], 'primary'>;
}

export const getMessagingIcon = (type: MessagingType, condensed?: boolean) => {
  const size = condensed ? 'large' : 'xLarge';

  switch (type) {
    case 'success':
      return <CheckCircleIcon color="success" size={size} />;

    case 'error':
      return <ErrorIcon color="danger" size={size} />;

    case 'warning':
      return <WarningIcon color="warning50" size={size} />;

    case 'info':
      return <InfoIcon color="primary60" size={size} />;
  }
};

export const getBorderStyle = (type: MessagingType, theme: ThemeInterface) => css`
  ${type === 'success' &&
  css`
    border-left: ${theme.spacing.xxSmall} solid ${theme.colors.success};
  `};

  ${type === 'error' &&
  css`
    border-left: ${theme.spacing.xxSmall} solid ${theme.colors.danger};
  `};

  ${type === 'warning' &&
  css`
    border-left: ${theme.spacing.xxSmall} solid ${theme.colors.warning50};
  `};

  ${type === 'info' &&
  css`
    border-left: ${theme.spacing.xxSmall} solid ${theme.colors.primary60};
  `};
`;

export const getActionVariant = (variant: ButtonProps['variant']) =>
  variant === 'primary' ? 'secondary' : variant;
