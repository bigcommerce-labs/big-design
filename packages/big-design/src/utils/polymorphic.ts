// Types from https://github.com/kripod/react-polymorphic-types with a few modifications for Big Design.

// Block external access to auxiliary types
export {};

type Merge<T, U> = Omit<T, keyof U> & U;

type PropsWithAs<
  P,
  T extends React.ElementType,
  S extends keyof JSX.IntrinsicElements = keyof JSX.IntrinsicElements,
> = P & {
  as?: T extends keyof JSX.IntrinsicElements ? (T extends S ? T : never) : never;
};

export type PolymorphicPropsWithoutRef<
  P,
  T extends React.ElementType,
  S extends keyof JSX.IntrinsicElements = keyof JSX.IntrinsicElements,
> = Merge<
  T extends keyof JSX.IntrinsicElements
    ? React.PropsWithoutRef<JSX.IntrinsicElements[T]>
    : React.ComponentPropsWithoutRef<T>,
  PropsWithAs<P, T, S>
>;

export type PolymorphicPropsWithRef<
  P,
  T extends React.ElementType,
  S extends keyof JSX.IntrinsicElements = keyof JSX.IntrinsicElements,
> = Merge<
  T extends keyof JSX.IntrinsicElements
    ? React.PropsWithRef<JSX.IntrinsicElements[T]>
    : React.ComponentPropsWithRef<T>,
  PropsWithAs<P, T, S>
>;

type PolymorphicExoticComponent<
  P = Record<string, unknown>,
  T extends React.ElementType = React.ElementType,
  S extends keyof JSX.IntrinsicElements = keyof JSX.IntrinsicElements,
> = Merge<
  React.ExoticComponent<P & { [key: string]: unknown }>,
  /**
   * **NOTE**: Exotic components are not callable.
   */
  <InstanceT extends React.ElementType = T>(
    props: PolymorphicPropsWithRef<P, InstanceT, S>,
  ) => React.ReactElement | null
>;

export type PolymorphicForwardRefExoticComponent<
  P,
  T extends React.ElementType,
  S extends keyof JSX.IntrinsicElements = keyof JSX.IntrinsicElements,
> = Merge<
  React.ForwardRefExoticComponent<P & { [key: string]: unknown }>,
  PolymorphicExoticComponent<P, T, S>
>;

export type PolymorphicMemoExoticComponent<
  P,
  T extends React.ElementType,
  S extends keyof JSX.IntrinsicElements = keyof JSX.IntrinsicElements,
> = Merge<React.MemoExoticComponent<React.ComponentType<any>>, PolymorphicExoticComponent<P, T, S>>;
