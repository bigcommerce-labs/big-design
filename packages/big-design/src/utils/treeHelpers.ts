import { TreeNodeProps, TreeProps } from '../components/Tree';

export function depthFirstSearch<T>(
  nodes: TreeProps<T>['initialNodes'],
  predicate: (node: TreeNodeProps<T>) => boolean,
  greedy?: true,
): TreeNodeProps<T> | null;
export function depthFirstSearch<T>(
  nodes: TreeProps<T>['initialNodes'],
  predicate: (node: TreeNodeProps<T>) => boolean,
  greedy?: false,
): TreeNodeProps<T>[] | null;
export function depthFirstSearch<T>(
  nodes: TreeProps<T>['initialNodes'],
  predicate: (node: TreeNodeProps<T>) => boolean,
  greedy?: boolean,
): TreeNodeProps<T>[] | TreeNodeProps<T> | null;
export function depthFirstSearch<T>(
  nodes: TreeProps<T>['initialNodes'],
  predicate: (node: TreeNodeProps<T>) => boolean,
  greedy = true,
): TreeNodeProps<T>[] | TreeNodeProps<T> | null {
  let list: TreeNodeProps<T>[] = [];

  for (const node of nodes) {
    if (predicate(node)) {
      if (greedy) {
        return node;
      }

      list.push(node);
    }

    if (node.children && node.children.length > 0) {
      if (greedy) {
        return depthFirstSearch(node.children, predicate, greedy);
      }

      const nestedDfs = depthFirstSearch(node.children, predicate, greedy);

      if (nestedDfs) {
        list = list.concat(nestedDfs);
      }
    }
  }

  return greedy ? null : list;
}
