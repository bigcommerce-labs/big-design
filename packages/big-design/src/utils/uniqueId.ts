let counter = 0;

// TODO: Remove once all components use useUniqueId:
export function uniqueId(prefix: string) {
  counter += 1;

  const id = counter;

  return prefix + id;
}
