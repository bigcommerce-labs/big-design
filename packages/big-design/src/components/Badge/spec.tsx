import { theme } from '@bigcommerce/big-design-theme';
import React from 'react';
import 'jest-styled-components';

import { render } from '@test/utils';

import { Badge } from './index';

test('has margin props', () => {
  const { container, rerender } = render(<Badge label="Badge" />);

  expect(container.firstChild).not.toHaveStyle('margin: 1rem');

  rerender(<Badge label="Badge" margin="medium" />);

  expect(container.firstChild).toHaveStyle('margin: 1rem');
});

test('render default Badge', () => {
  const { container } = render(<Badge label="Badge" />);

  expect(container.firstChild).toMatchSnapshot();
  expect(container.firstChild).toHaveStyle(`background-color: ${theme.colors.secondary60}`);
});

test('render success Badge', () => {
  const { container } = render(<Badge label="Badge" variant="success" />);

  expect(container.firstChild).toMatchSnapshot();
  expect(container.firstChild).toHaveStyle(`background-color: ${theme.colors.success50}`);
});

test('render danger Badge', () => {
  const { container } = render(<Badge label="Badge" variant="danger" />);

  expect(container.firstChild).toMatchSnapshot();
  expect(container.firstChild).toHaveStyle(`background-color: ${theme.colors.danger40}`);
});

test('render warning Badge', () => {
  const { container } = render(<Badge label="Badge" variant="warning" />);

  expect(container.firstChild).toMatchSnapshot();
  expect(container.firstChild).toHaveStyle(`background-color: ${theme.colors.warning40}`);
});

test('render secondary Badge', () => {
  const { container } = render(<Badge label="Badge" variant="secondary" />);

  expect(container.firstChild).toMatchSnapshot();
  expect(container.firstChild).toHaveStyle(`background-color: ${theme.colors.secondary60}`);
});

test("doesn't render if label prop is invalid", () => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
  // @ts-ignore
  const { container } = render(<Badge label={<p>Label</p>} />);

  expect(container.firstChild).toBe(null);
});
