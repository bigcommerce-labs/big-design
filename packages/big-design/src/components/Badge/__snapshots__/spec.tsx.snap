// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render danger Badge 1`] = `
.c0 {
  color: #FFFFFF;
  border-radius: 0.25rem;
  display: inline-block;
  font-size: 0.75rem;
  font-weight: 600;
  text-align: center;
  text-transform: uppercase;
  padding: 0.25rem 0.5rem;
  background-color: #DB3643;
}

<span
  class="c0"
>
  Badge
</span>
`;

exports[`render default Badge 1`] = `
.c0 {
  color: #FFFFFF;
  border-radius: 0.25rem;
  display: inline-block;
  font-size: 0.75rem;
  font-weight: 600;
  text-align: center;
  text-transform: uppercase;
  padding: 0.25rem 0.5rem;
  background-color: #5E637A;
}

<span
  class="c0"
>
  Badge
</span>
`;

exports[`render secondary Badge 1`] = `
.c0 {
  color: #FFFFFF;
  border-radius: 0.25rem;
  display: inline-block;
  font-size: 0.75rem;
  font-weight: 600;
  text-align: center;
  text-transform: uppercase;
  padding: 0.25rem 0.5rem;
  background-color: #5E637A;
}

<span
  class="c0"
>
  Badge
</span>
`;

exports[`render success Badge 1`] = `
.c0 {
  color: #FFFFFF;
  border-radius: 0.25rem;
  display: inline-block;
  font-size: 0.75rem;
  font-weight: 600;
  text-align: center;
  text-transform: uppercase;
  padding: 0.25rem 0.5rem;
  background-color: #208831;
}

<span
  class="c0"
>
  Badge
</span>
`;

exports[`render warning Badge 1`] = `
.c0 {
  color: #FFFFFF;
  border-radius: 0.25rem;
  display: inline-block;
  font-size: 0.75rem;
  font-weight: 600;
  text-align: center;
  text-transform: uppercase;
  padding: 0.25rem 0.5rem;
  color: #313440;
  background-color: #FFBF00;
}

<span
  class="c0"
>
  Badge
</span>
`;
